src/libstd/io/mod.rs

```diff
@@ -1549,93 +1594,182 @@ impl<R: Read> Iterator for Bytes<R> {
+        macro_rules! read_byte {
+            (EOF => $on_eof: expr) => {
+                {
+                    let byte;
+                    loop {
+                        match self.inner.fill_buf() {
+                            Ok(buffer) => {
+                                if let Some(&b) = buffer.first() {
+                                    byte = b;
+                                    break
+                                } else {
+                                    $on_eof
+                                }
+                            }
+                            Err(ref e) if e.kind() == ErrorKind::Interrupted => {}
+                            Err(e) => return Some(Err(Utf8CharsError::Io(e))),
+                        }
+                    }
+                    byte
+                }
+            }
+        }
 
-    fn next(&mut self) -> Option<result::Result<char, CharsError>> {
-        let mut buf = [0];
-        let first_byte = match self.inner.read(&mut buf) {
-            Ok(0) => return None,
-            Ok(..) => buf[0],
-            Err(e) => return Some(Err(CharsError::Other(e))),
-        };
-        let width = core_str::utf8_char_width(first_byte);
-        if width == 1 { return Some(Ok(first_byte as char)) }
-        if width == 0 { return Some(Err(CharsError::NotUtf8)) }
-        let mut buf = [first_byte, 0, 0, 0];
-        {
-            let mut start = 1;
-            while start < width {
-                match self.inner.read(&mut buf[start..width]) {
-                    Ok(0) => return Some(Err(CharsError::NotUtf8)),
-                    Ok(n) => start += n,
-                    Err(e) => return Some(Err(CharsError::Other(e))),
+        let first = read_byte!(EOF => return None);
+        self.inner.consume(1);
+
+        macro_rules! continuation_byte {
+            ($range: pat) => {
+                {
+                    match read_byte!(EOF => return Some(Err(Utf8CharsError::IncompleteUtf8))) {
+                        byte @ $range => {
+                            self.inner.consume(1);
+                            (byte & 0b0011_1111) as u32
+                        }
+                        _ => return Some(Err(Utf8CharsError::InvalidUtf8))
+                    }
                 }
             }
         }
-        Some(match str::from_utf8(&buf[..width]).ok() {
-            Some(s) => Ok(s.chars().next().unwrap()),
-            None => Err(CharsError::NotUtf8),
-        })
+
+        // Ranges can be checked against https://tools.ietf.org/html/rfc3629#section-4
+        let code_point = match core_str::utf8_char_width(first) {
+            1 => return Some(Ok(first as char)),
+            2 => {
+                let second = continuation_byte!(0x80...0xBF);
+                ((first & 0b0001_1111) as u32) << 6 | second
+            }
+            3 => {
+                let second = match first {
+                    0xE0        => continuation_byte!(0xA0...0xBF),
```

Comment:
I don’t understand what you mean by substitute.

http://www.cl.cam.ac.uk/%7Emgk25/ucs/examples/UTF-8-test.txt doesn’t really have test cases:
> This file is not meant to be a conformance test. It does not prescribe any particular outcome. Therefore, there is no way to "pass" or "fail" this test file